File: .fieldwork-authority-deny-scope.py
from pathlib import Path

path = Path("crates/tauri/src/ipc/authority.rs")
text = path.read_text()


def replace_once(old: str, new: str, label: str) -> None:
    global text
    count = text.count(old)
    if count != 1:
        raise SystemExit(f"{label}: expected exactly one source match, found {count}")
    text = text.replace(old, new, 1)


origin_impl = '''impl Origin {
  fn matches(&self, context: &ExecutionContext) -> bool {
    match (self, context) {
      (Self::Local, ExecutionContext::Local) => true,
      (Self::Remote { url }, ExecutionContext::Remote { url: url_pattern }) => {
        url_pattern.test(url)
      }
      _ => false,
    }
  }
}
'''

replace_once(
    origin_impl,
    origin_impl
    + '''
fn resolved_command_matches(
  command: &ResolvedCommand,
  window: &str,
  webview: &str,
  origin: &Origin,
) -> bool {
  origin.matches(&command.context)
    && (command.webviews.iter().any(|w| w.matches(webview))
      || command.windows.iter().any(|w| w.matches(window)))
}
''',
    "shared matcher insertion",
)

old_debug = '''    if let Some(resolved) = self.denied_commands.get(&command) {
      format!(
        "{command_pretty_name} explicitly denied on origin {origin}\\n\\nreferenced by: {}",
        print_references(resolved)
      )
    } else {
      let command_matches = self.allowed_commands.get(&command);

      if let Some(resolved) = self.allowed_commands.get(&command) {
        let resolved_matching_origin = resolved
          .iter()
          .filter(|cmd| origin.matches(&cmd.context))
          .collect::<Vec<&ResolvedCommand>>();
        if resolved_matching_origin
          .iter()
          .any(|cmd| cmd.webviews.iter().any(|w| w.matches(webview)))
          || resolved_matching_origin
            .iter()
            .any(|cmd| cmd.windows.iter().any(|w| w.matches(window)))
        {
          "allowed".to_string()
'''

new_debug = '''    let matching_denied = self.denied_commands.get(&command).and_then(|resolved| {
      let matching = resolved
        .iter()
        .filter(|cmd| resolved_command_matches(cmd, window, webview, origin))
        .cloned()
        .collect::<Vec<_>>();
      (!matching.is_empty()).then_some(matching)
    });

    if let Some(resolved) = matching_denied {
      format!(
        "{command_pretty_name} explicitly denied on window \\"{window}\\", webview \\"{webview}\\", origin {origin}\\n\\nreferenced by: {}",
        print_references(&resolved)
      )
    } else {
      let command_matches = self.allowed_commands.get(&command);

      if let Some(resolved) = self.allowed_commands.get(&command) {
        if resolved
          .iter()
          .any(|cmd| resolved_command_matches(cmd, window, webview, origin))
        {
          "allowed".to_string()
'''

replace_once(old_debug, new_debug, "debug access matcher")

old_resolve = '''    if self
      .denied_commands
      .get(command)
      .map(|resolved| resolved.iter().any(|cmd| origin.matches(&cmd.context)))
      .is_some()
    {
      None
    } else {
      self.allowed_commands.get(command).and_then(|resolved| {
        let resolved_cmds = resolved
          .iter()
          .filter(|cmd| {
            origin.matches(&cmd.context)
              && (cmd.webviews.iter().any(|w| w.matches(webview))
                || cmd.windows.iter().any(|w| w.matches(window)))
          })
          .cloned()
          .collect::<Vec<_>>();
'''

new_resolve = '''    if self
      .denied_commands
      .get(command)
      .is_some_and(|resolved| {
        resolved
          .iter()
          .any(|cmd| resolved_command_matches(cmd, window, webview, origin))
      })
    {
      None
    } else {
      self.allowed_commands.get(command).and_then(|resolved| {
        let resolved_cmds = resolved
          .iter()
          .filter(|cmd| resolved_command_matches(cmd, window, webview, origin))
          .cloned()
          .collect::<Vec<_>>();
'''

replace_once(old_resolve, new_resolve, "runtime access matcher")

path.write_text(text)

File: crates/tauri/tests/fieldwork_authority_deny_scope.rs
// Copyright 2019-2024 Tauri Programme within The Commons Conservancy
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

use std::collections::BTreeMap;

use tauri::ipc::{Origin, RuntimeAuthority};
use tauri_utils::acl::{
  resolved::{Resolved, ResolvedCommand},
  ExecutionContext,
};

fn command(context: ExecutionContext, windows: &[&str], webviews: &[&str]) -> ResolvedCommand {
  ResolvedCommand::new(
    context,
    Default::default(),
    windows
      .iter()
      .map(|pattern| pattern.parse().unwrap())
      .collect(),
    webviews
      .iter()
      .map(|pattern| pattern.parse().unwrap())
      .collect(),
    None,
  )
}

fn authority(allowed: Vec<ResolvedCommand>, denied: Vec<ResolvedCommand>) -> RuntimeAuthority {
  let mut allowed_commands = BTreeMap::new();
  allowed_commands.insert("fieldwork".to_string(), allowed);
  let mut denied_commands = BTreeMap::new();
  denied_commands.insert("fieldwork".to_string(), denied);

  RuntimeAuthority::new(
    BTreeMap::new(),
    Resolved {
      allowed_commands,
      denied_commands,
      ..Default::default()
    },
  )
}

#[test]
fn deny_for_other_origin_does_not_block_local_access() {
  let authority = authority(
    vec![command(ExecutionContext::Local, &["main"], &[])],
    vec![command(
      ExecutionContext::Remote {
        url: "https://denied.example/*".parse().unwrap(),
      },
      &["main"],
      &[],
    )],
  );

  assert!(
    authority
      .resolve_access("fieldwork", "main", "main", &Origin::Local)
      .is_some(),
    "deny rule for another origin blocked local access"
  );
}

#[test]
fn deny_for_other_window_does_not_block_main_window() {
  let authority = authority(
    vec![command(ExecutionContext::Local, &["main"], &[])],
    vec![command(ExecutionContext::Local, &["admin"], &[])],
  );

  assert!(
    authority
      .resolve_access("fieldwork", "main", "main", &Origin::Local)
      .is_some(),
    "deny rule for another window blocked main-window access"
  );
}

#[test]
fn matching_deny_still_overrides_matching_allow() {
  let authority = authority(
    vec![command(ExecutionContext::Local, &["main"], &[])],
    vec![command(ExecutionContext::Local, &["main"], &[])],
  );

  assert!(
    authority
      .resolve_access("fieldwork", "main", "main", &Origin::Local)
      .is_none(),
    "matching deny did not override matching allow"
  );
}
